### PR TITLE
Switch PLISTS and WeakPointerObjs to use IntObj to store the size

### PR DESCRIPTION
**Pull request by ChrisJefferson**
This switches PLISTS and WeakPointerObjs to use an IntObj to store their size.

This is part of a cleanup towards making all the builtin data types use IntObj to store their size.

The only notable problem is that now a compeletely zeroed block of memory isn't a valid Plist, as it needs to store INTOBJ_INT(0) as it's size.

There are two reasons to do this:

* It helps make clearer these datastructures can only be as big as an intobj (which was already true anyway in practice)
* It is required to get rid of a bug in saveworkspace (in a neat way; see #3505).

### CODE REVIEW

**Comment by ChrisJefferson** [comment]
Note that I have some tests to ensure that these behave well, and there are no problems are 2^28, but I'm not sure I want to commit them and have them regularly tested, as they can only be run on 32-bit machines, needs lots of memory, and also they require running lots of small test files in seperate GAPs, as sometimes GASMAN fails to collect the huge bags I'm making and GAP runs out of memory.

**Comment by fingolfin** [comment]
I just grepped through packages for `NewBag.*T_PLIST` and it seems none of them do that (unless they put a newline after `NewBag` and before `T_PLIST`... but then, shame on them for producing unreadable code). Good! I'll properly review this PR after I am done teaching.

File: src/hpc/threadapi.c [hunk 1]
```diff
@@ -374,10 +374,10 @@ static pthread_mutex_t KeepAliveLock;
 
 Obj KeepAlive(Obj obj)
 {
-    Obj newKeepAlive = NewBag(T_PLIST, 4 * sizeof(Obj));
+    Obj newKeepAlive = NEW_PLIST(T_PLIST, 4);
```
Comment (fingolfin) [inline src/hpc/threadapi.c]: I like this. I think we should actually add even more `NEW_FOO` functions, and try to call `NewBag` directly as little as possible.

File: src/weakptr.c [hunk 2]
```diff
@@ -636,7 +636,7 @@ static Obj CopyObjWPObj(Obj obj, Int mut)
         ADDR_OBJ(copy)[0] = CONST_ADDR_OBJ(obj)[0];
     }
     else {
-        copy = NewBag( T_PLIST+IMMUTABLE, SIZE_OBJ(obj) );
+        copy = NEW_PLIST( T_PLIST+IMMUTABLE, LengthWPObj(obj) );
         SET_LEN_PLIST(copy,LengthWPObj(obj));
```
Comment (fingolfin) [inline src/weakptr.c]: Note that the value of `LengthWPObj(obj)` could change between the two previous lines. And the loop a few lines later again uses `SIZE_OBJ(obj)`...

File: src/weakptr.c [hunk 3]
```diff
@@ -636,7 +645,7 @@ static Obj CopyObjWPObj(Obj obj, Int mut)
         ADDR_OBJ(copy)[0] = CONST_ADDR_OBJ(obj)[0];
     }
     else {
-        copy = NewBag( T_PLIST+IMMUTABLE, SIZE_OBJ(obj) );
+        copy = NEW_PLIST(T_PLIST + IMMUTABLE, LengthWPObj(obj));
```
Comment (fingolfin) [inline src/weakptr.c]:
```suggestion
        copy = NEW_PLIST_IMM(T_PLIST, LengthWPObj(obj));
```
I was trying in the past to get rid of as many uses of `IMMUTABLE` as possible, in the hope that at some point we turn it into an object flag (also Markus worked hard towards that goal).

**Pull request merged**